evalml/pipelines/time_series_pipeline_base.py

```diff
@@ -67,11 +69,161 @@ def fit(self, X, y):
         self._fit(X, y)
         return self
 
+    @staticmethod
+    def _move_index_forward(index, gap):
+        """Fill in the index of the gap features and values with the right values."""
+        if isinstance(index, (pd.DatetimeIndex, pd.PeriodIndex, pd.TimedeltaIndex)):
+            return index.shift(gap)
+        else:
+            return index + gap
+
+    @staticmethod
+    def _are_datasets_separated_by_gap(train_index, test_index, gap):
+        """Determine if the train and test datasets are separated by gap number of units.
+
+        This will be true when users are predicting on unseen data but not during cross
+        validation since the target is known.
+        """
+        gap_difference = gap + 1
+        index_difference = test_index[0] - train_index[-1]
+        if isinstance(
+            train_index, (pd.DatetimeIndex, pd.PeriodIndex, pd.TimedeltaIndex)
+        ):
+            gap_difference *= test_index.freq
```

Comment:
This should work regardless of whether `date_index` is `None` or not since this logic only relies on the data being ordered sequentially. To your point, I added some unit tests to `predict` and `predict_in_sample` where `date_index` is not `None` and it still works. 

So I think this is sufficient for a first cut but I am totally open to reworking this in the future. I think it may indeed be simpler to just rely on `date_index`.